testing/integration/upgrade_test.go

```diff
@@ -899,3 +899,89 @@ func removePackageVersionFiles(t *testing.T, f *atesting.Fixture) {
 		require.NoErrorf(t, err, "error removing package version file %q", vFile)
 	}
 }
+
+// TestStandaloneUpgradeFailsStatus tests the scenario where upgrading to a new version
+// of Agent fails due to the new Agent binary reporting an unhealthy status. It checks
+// that the Agent is rolled back to the previous version.
+func TestStandaloneUpgradeFailsStatus(t *testing.T) {
+	define.Require(t, define.Requirements{
+		Local:   false, // requires Agent installation
+		Isolate: false,
+		Sudo:    true, // requires Agent installation
+	})
+
+	upgradeFromVersion, err := version.ParseVersion(define.Version())
+	require.NoError(t, err)
+
+	// We go back TWO minors because sometimes we are in a situation where
+	// the current version has been advanced to the next release (e.g. 8.10.0)
+	// but the version before that (e.g. 8.9.0) hasn't been released yet.
+	previousVersion, err := upgradeFromVersion.GetPreviousMinor()
+	require.NoError(t, err)
+	previousVersion, err = previousVersion.GetPreviousMinor()
+	require.NoError(t, err)
+
+	// For testing the upgrade we actually perform a downgrade
+	upgradeToVersion := previousVersion
+
+	t.Logf("Testing Elastic Agent upgrade from %s to %s...", upgradeFromVersion, upgradeToVersion)
+
+	agentFixture, err := define.NewFixture(t, define.Version())
+	require.NoError(t, err)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	err = agentFixture.Prepare(ctx)
+	require.NoError(t, err, "error preparing agent fixture")
+
+	// Configure Agent with fast watcher configuration and also an invalid
+	// input when the Agent version matches the upgraded Agent version. This way
+	// the pre-upgrade version of the Agent runs healthy, but the post-upgrade
+	// version doesn't.
+	invalidInputPolicy := fastWatcherCfg + fmt.Sprintf(`
```

Comment:
How do we guarantee that the fastWatcherCfg is supported? is it even needed here ?

Reply:
We can't guarantee it but the settings in `fastWatcherCfg` should just be ignored by any versions of Agent that don't support it.  Also, `checkUpgradeWatcherRan` will just skip checking if the upgrade watcher finished running for Agents that don't support the `fastWatcherCfg` settings: https://github.com/elastic/elastic-agent/blob/22c276ddc2241bb73322c7968edef9a9d7091f43/testing/integration/upgrade_test.go#L558-L565

So I think we are okay here?

Reply:
If  the checkUpgradeWatcherRan skips here, doesn't that make the test run pointless?
If I understood correctly, the point of the test is to verify that the watcher rollback the upgrade because of the unhealthy state of the new agent (please correct me if this is not the correct assumption). In such a case wouldn't the upgrade watcher terminate before the full 10 mins default grace period because of the amount of errors ?

Reply:
I think that the check is correct as it refers to the earliest version that understand settings for a shorter watch period... The fact that the watcher is spawned from the old or the new shouldn't matter as the outcome in the case of an unhealthy upgraded agent should stay the same...
It actually provides an interesting additional test: that even if the upgraded agent process itself is crashing, the agent process stays alive and take the correct action ;)